pkg/utils/metadata.go

```diff
@@ -44,6 +53,22 @@ func ProjectID(project string, client metadataClient.Client) (string, error) {
 	return projectID, nil
 }
 
+// ProjectIDOrDefault returns the project ID by performing the following order:
+// 1) if the input project ID is valid, simply use it.
+// 2) if there is a PROJECT_ID environmental variable, use it.
+// 3) use metadataClient to resolve project id.
+func ProjectIDOrDefault(ctx context.Context, projectID string) (string, error) {
+	if projectID != "" {
+		return projectID, nil
+	}
+	// Parse project id from env var, if not found we simply continue
+	var env ProjectIDEnvConfig
```

Comment:
I am not sure if this makes things more complicated. `envConfig` is used in lots of places. In most of the cases, there are multiple fields in `envConfig` struct (in addition to `ProjectID`), and the common way is using `envconfig.Process` to parse all the fields. Here, you picked up two cases whose `envConfig` happen to only have `ProjectID`, and wrapped the `envconfig.Process` in new `ProjectIDOrDefault`, what if later those two places need more fields in `envConfig` struct (may cause duplicate usage of `envconfig.Process`)?

Reply:
Yes, I am aware of that and have thought about this very carefully. If you see my previous cleanup for broker and trigger, I kept the parsing part in the controller because of exactly what you said. Here is the reason I decided to do this now:

1. This code is not performance sensitive. And the the performance of `Process` function is similar to `os.lookupEnv` IIUC.
2. In case we want to pass the env var, very likely we will save the result, like what we do in the broker controller.
3. I am happy to use `os.Getenv("PROJECT_ID")` here if we are really concerned about performance difference, but from last conversation with @grantr envconfig a preferred way.

Reply:
For me, it is less readable to change like this. Before this change, developers only need to know `envconfig.Process` logic in the files which has `envConfig` struct. Now, since you are changing it in all parts (include places where doesn’t need to deal with `envconfig` struct), developers need to know this parsing logic due to the usage of `ProjectIDOrDefault` function. It's an overexposure in my understanding. But that’s just my personal opinion.

Reply:
I am actually not sure what you mean here. User will never need to use `envconfig.Process` or even import `envconfig` directly after this. The only function exposed here is `ProjectIDOrDefault`. In fact, I can simply use `os.Getenv` here, but either way user should not worry about it at all. Do I miss sth here?

Reply:
I mean when developers use `ProjectIDOrDefault`, they will likely look into the logic, and found it is checking `envconfig`, if they don't know the detail of parsing logic for it (suppose they are not changing a file which has `envconfig`), they need to have additional knowledge to know whether this `ProjectIDOrDefault` is suitable for their use case. It increases cognitive load for them. It might be better to have a clear function name (as Cong suggested) and together with more clear comments.

Reply:
Sorry about the confusion here. This PR is not for clean up. It is mainly to introduce env var to sources #1781 , or in another word it is to have one single way of resolving project id in `knative-gcp`. The idea is that user shouldn't know the implementation details, or change their code when we decide to change the logic of resolving project id. And caller shouldn't need to worry about the details, as the function is used as a golden way to resolve project id in `knative-gcp`.

Reply:
Though I agree that different code can have different ways to resolve project id differently if it is on purpose, from my previous discussion with Grant and Nacho in `knative-gcp` the difference is more due to inconsistency of the code. Eg, in broker, why would controller consider env var but not in the reconciler? Same thing goes to sources, env var supported was added to brokers after sources, and the logic has got out of sync. So I am trying to have a golden way of resolving project id if not provided that we shouldn't need to worry about the implementation details at all.

Reply:
We now used `os.Getenv` in `init` function.